### PR TITLE
[Indexing] Add transform from scatter/gather to insert_slice/extract_slice

### PR DESCRIPTION
**Pull request by makslevental**
Depends on https://github.com/iree-org/iree-llvm-sandbox/pull/753

This PR adds a lowering from the simplest scatter/gather case (regular start, stop, stop for all indices) to `tensor.insert_slice`/`tensor.extract_slice`.

### CODE REVIEW

File: include/structured/Dialect/Indexing/IR/IndexingOps.td [hunk 1]
```diff
@@ -50,16 +55,21 @@ def Indexing_ScatterOp : Indexing_Op<"scatter", [
...
+    $source `into` $dest `[` $indices `]`
+            `scatter_dims` `(` $scatter_dims `)`
+            (`unique` $unique^)?
+            (`is_slice` $is_slice^)?
+            attr-dict `:` functional-type(operands, results)
   }];
   let extraClassDeclaration = [{
     static bool isCompatibleReturnTypes(TypeRange l, TypeRange r);
   }];
+  // let hasCanonicalizer = 1;
```
Comment (nicolasvasilache) [inline include/structured/Dialect/Indexing/IR/IndexingOps.td:72]: dead code

File: test/python/dialects/indexing/dialect.py [hunk 2]
```diff
@@ -1306,15 +1306,88 @@ def testDynSliceScatter():
...
+  # CHECK:     %2 = tensor.empty() : tensor<10x30xf32>
+  # CHECK:     %3 = scf.for %arg0 = %c0 to %c10 step %c5 iter_args(%arg1 = %2) -> (tensor<10x30xf32>) {
+  # CHECK:       %4 = scf.for %arg2 = %c0 to %c30 step %c5 iter_args(%arg3 = %arg1) -> (tensor<10x30xf32>) {
+  # CHECK:         %5 = scf.for %arg4 = %c0 to %c20 step %c5 iter_args(%arg5 = %arg3) -> (tensor<10x30xf32>) {
+  # CHECK:           %extracted_slice = tensor.extract_slice %0[%arg0, %arg4] [%c5, %c5] [%c1, %c1] : tensor<10x20xf32> to tensor<?x?xf32>
+  # CHECK:           %extracted_slice_0 = tensor.extract_slice %1[%arg4, %arg2] [%c5, %c5] [%c1, %c1] : tensor<20x30xf32> to tensor<?x?xf32>
+  # CHECK:           %extracted_slice_1 = tensor.extract_slice %arg5[%arg0, %arg2] [%c5, %c5] [%c1, %c1] : tensor<10x30xf32> to tensor<?x?xf32>
+  # CHECK:           %6 = tensor.empty(%c5, %c5) : tensor<?x?xf32>
+  # CHECK:           %7 = linalg.fill ins(%cst : f32) outs(%6 : tensor<?x?xf32>) -> tensor<?x?xf32>
+  # CHECK:           %8 = linalg.matmul {cast = #linalg.type_fn<cast_signed>} ins(%extracted_slice, %extracted_slice_0 : tensor<?x?xf32>, tensor<?x?xf32>) outs(%7 : tensor<?x?xf32>) -> tensor<?x?xf32>
```
Comment (nicolasvasilache) [inline test/python/dialects/indexing/dialect.py:1381]: This should take extracted_slice_1 as the outs parameter and then you don't need the extra empty, fill and add.

File: include/structured/Dialect/Indexing/IR/IndexingOps.td [hunk 3]
```diff
@@ -29,16 +29,21 @@ def Indexing_GatherOp : Indexing_Op<"gather", [
   let arguments = (ins AnyRankedTensor:$source,
                        RankedTensorOf<[AnySignlessIntegerOrIndex]>:$indices,
                        DenseI64ArrayAttr:$gather_dims,
-                       UnitAttr:$unique);
+                       UnitAttr:$unique,
+                       UnitAttr:$is_slice);
```
Comment (nicolasvasilache) [inline include/structured/Dialect/Indexing/IR/IndexingOps.td:33]: This is quite dangerous to implement as an attribute and it should be very easy to write a proper helper to detect that your indices come from some arange.

**Review by nicolasvasilache — CHANGES_REQUESTED** [review]

File: test/python/dialects/indexing/dialect.py (hunk 2 above)
Comment (makslevental) [reply test/python/dialects/indexing/dialect.py]: add such a test case

Comment (makslevental) [reply test/python/dialects/indexing/dialect.py]: i take it back - maybe it's not so questionable after all :p (implemented `c += a @ b` with no fill.

**Pull request closed without merging**